packages/typegpu/src/core/function/ioOutputStruct.ts

```diff
@@ -0,0 +1,65 @@
+import { isBuiltin, struct } from '../../data';
+import { attribute, location } from '../../data/attributes';
+import { getCustomLocation, isData } from '../../data/dataTypes';
+import type { BaseWgslData, WgslStruct } from '../../data/wgslTypes';
+import type { FnCore } from './fnCore';
+import type { IOData, IOLayout, Implementation } from './fnTypes';
+
+export type IOLayoutToOutputStruct<T extends IOLayout> = T extends BaseWgslData
+  ? WgslStruct<{ out: T }>
+  : T extends Record<string, BaseWgslData>
+    ? WgslStruct<T>
+    : never;
+
+export function withLocations(
+  members: Record<string, BaseWgslData>,
+): Record<string, BaseWgslData> {
+  let nextLocation = 0;
+
+  return Object.fromEntries(
+    Object.entries(members).map(([key, member]) => {
+      if (isBuiltin(member)) {
+        // Skipping builtins
+        return [key, member];
+      }
+
+      const customLocation = getCustomLocation(member);
+      if (customLocation !== undefined) {
+        // This member is already marked, start counting from the next location over.
+        nextLocation = customLocation + 1;
+        return [key, member];
+      }
+
+      return [
+        key,
+        attribute(member, { type: '@location', value: nextLocation++ }),
+      ];
+    }),
+  );
+}
+
+export function createOutputStruct<T extends IOData>(
+  core: FnCore,
+  implementation: Implementation,
+  returnType: IOLayout<T>,
+) {
+  const Output = struct(
+    withLocations(
+      isData(returnType) ? { out: location(0, returnType) } : returnType,
+    ) as Record<string, T>,
+  );
+
+  if (typeof implementation === 'string') {
+    const outputName = implementation
+      .match(/->(?<output>.*?){/s)
+      ?.groups?.output?.trim();
+
+    if (outputName) {
+      const externals: Record<string, unknown> = {};
+      externals[outputName] = Output;
+      core.applyExternals(externals);
+    }
+  }
```

Comment:
What about the situation when we have for example:
```wgsl
@vertex 
fn main(@builtin(vertex_index) vertexIndex: u32) -> @builtin(position) vec4f {
   // ....
}
```

Reply:
there will be one unnecessary string replace operation, but I think we can live with that

Reply:
in this case it adds to uses { `@builtin(position) vec4f`: Output}, then tries to replace `@builtin(position) vec4f` with `vertex_fn_Output_1` in resolved code, but it doesn't work, I assume because of the parentheses, as they aren't escaped in the regex